---
title: "Creating Tx2Gene file"
author: "Jonathan Anzules"
date: "2023-09-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(tximport) # Importing quant_sf files
library(GenomicFeatures) #Useful genomics tools
library(BiocManager)
library(rhdf5) #To make an hdf5 files
library(Rtsne)
library(DESeq2) # For normalizing
library(scran) # normalization based on cell clusters
```

**To Do**
- When I move the hdf5 file add something that also moves or makes a metadata file of what information is contained
- Learn how to work with the txi data
- cluster the cells and visualize with tsne and Umap
  - Identify the which genes are the most prevalent in each cluster.
- Perform a pseudo time analysis



```{r}
# txdb <- makeTxDbFromGFF("/Documents and Settings/jonan/Documents/Genomes/Homo-Sapiens/GRCh38/gencode.v44.chr_patch_hapl_scaff.annotation.gtf.gz")
# 
# # keytypes(txdb)
# k <- keys(txdb, keytype = "TXNAME")
# 
# tx2gene <- select(txdb, k, "GENEID", "TXNAME") # only works om 4.3.0 # Restart and clear output if this doesn't work
# # I think code is getting confused with another select() in another package, I'm not sure which one it is yet.
# tx2gene <- as.data.frame(tx2gene)
# tx2gene <- tx2gene[, c("TXNAME", "GENEID")]
# 
# #Saving the Dataframe
# write.csv(tx2gene, "/Documents and Settings/jonan/Documents/Genomes/Homo-Sapiens/GRCh38/tx2gene-Mappings", row.names=FALSE)
```

```{r - Preparing quantfiles}
#Grabbing the quant.sf files metadata and editing paths to work with R
T2D_files <- read.csv("/Documents and Settings/jonan/Documents/1Work/scWork/Data/PANCDB/Cleaned-N-Aligned-hpapdata/metadata_T2D.csv")
T2D_files$quant_file <- gsub("^/mnt/c", "C:", T2D_files$quant_file)

# dim(T2D_files)

#Reading in the tx2gene dataframe
tx2gene <- read.csv("/Documents and Settings/jonan/Documents/Genomes/Homo-Sapiens/GRCh38/tx2gene-Mappings")
```

```{r - Removing files that have failed}
# Grabbing files address
files <- T2D_files$quant_file

missing_files <- files[!file.exists(files)]

# Removing missing files from T2D_files dataframe
T2D_files <- T2D_files[!T2D_files$quant_file %in% missing_files,]

dim(T2D_files)
length(missing_files)
```

```{r - Creating the HDF5 file}

# Grabbing files address
files <- T2D_files$quant_file

txi <- tximport(files, type = "salmon", tx2gene = tx2gene)

# Assign cell IDs from metadata as column names of the count matrix
colnames(txi$counts) <- T2D_files$cell
str(txi)



txi$length
# # Now proceed to save the count matrix as an HDF5 file
h5createFile("/Documents and Settings/jonan/Documents/1Work/scWork/Data/PANCDB/Cleaned-N-Aligned-hpapdata/txi_data_T2D.h5")
h5write(txi$counts, "/Documents and Settings/jonan/Documents/1Work/scWork/Data/PANCDB/Cleaned-N-Aligned-hpapdata/txi_data.h5", "counts")
```

```{r}
# Create an HDF5 file
hdf5_filename <- "/Documents and Settings/jonan/Documents/1Work/scWork/Data/PANCDB/Cleaned-N-Aligned-hpapdata/txi_data_T2D.h5"
h5createFile(hdf5_filename)

# Store each component of the txi list in the HDF5 file
h5write(txi$abundance, hdf5_filename, "abundance")
h5write(txi$counts, hdf5_filename, "counts")
h5write(txi$length, hdf5_filename, "length")
h5write(txi$countsFromAbundance, hdf5_filename, "countsFromAbundance")
h5write(colnames(txi$counts), hdf5_filename, "counts_colnames")
h5write(rownames(txi$counts), hdf5_filename, "counts_rownames")

```
# Analyzing count matrix

```{r - Reading in the count matrix}

library(rhdf5)

# Read each component from the HDF5 file into a list
txi_read <- list(
  abundance = h5read(hdf5_filename, "abundance"),
  counts = h5read(hdf5_filename, "counts"),
  length = h5read(hdf5_filename, "length"),
  countsFromAbundance = h5read(hdf5_filename, "countsFromAbundance")
)

# Reading and assigning row and column names
rownames(txi_read$counts) <- h5read(hdf5_filename, "counts_rownames")
colnames(txi_read$counts) <- h5read(hdf5_filename, "counts_colnames")

str(txi_read)
```

```{r - Normalizing Data - Size factor estimates to clusters}

clusters <- quickCluster(txi$counts)

hist(as.numeric(clusters))
```


```{r}
pca_res <- prcomp(t(log1p(txi$counts)))
tsne_res <- Rtsne(pca_res$x[, 1:9], perplexity=2)

pca_res$x
tsne_res <- Rtsne(pca_res$x, perplexity=2)  # using first 10 PCA components
plot(tsne_res$Y, col = rainbow(nrow(txi$counts)), pch = 20, main = "t-SNE plot", xlab = "", ylab = "")

```
```{r}

# Run the t-SNE algorithm
tsne_res <- Rtsne(txi$counts, perplexity=30)  # adjust perplexity as needed

# Plot the results
plot(tsne_res$Y, t='n', main="t-SNE plot")
text(tsne_res$Y, labels=row.names(data_matrix), cex=0.7, col="blue")

```








